ykcapi/yk.h

```diff
@@ -28,8 +28,14 @@ typedef uint32_t YkHotThreshold;
 
 typedef struct YkMT YkMT;
 
-// Create a new `YkMT` instance.
-YkMT *yk_mt_new(void);
+// Create a new `YkMT` instance. If this fails then:
+//   * If `err_msg` is `NULL`, this function will abort.
+//   * If `err_msg` is not `NULL`:
+//       1. A malloc()d string with an error message explaining the failure
```

Comment:
Oh, yeah, this isn't right:

[`into_raw()` docs](https://doc.rust-lang.org/std/ffi/struct.CString.html#method.into_raw):

> The pointer which this function returns must be returned to Rust and reconstituted using [CString::from_raw](https://doc.rust-lang.org/std/ffi/struct.CString.html#method.from_raw) to be properly deallocated. Specifically, one should not use the standard C free() function to deallocate this string.

Reply:
We don't provide a way for C to deallocate the `CString` though...

Reply:
Hmm, yeah :/ That serves me right for not reading things properly! There might not be a good way of doing what I want because Rust and C might not share a malloc. Hmph. Will investigate later.

Reply:
It may simplify memory management to introduce an integer error code that can be stringified by an api call?

Reply:
Integer error codes are a maintenance disaster: they lock you in for all time and don't allow expansion.

Reply:
I don't understand how it's different to returning a malloced string in that regard?

Reply:
OK maybe I get it. The string can change freely, but when the API stabilises an integer has to have a stable meaning.

But imagine if libc working this way...

```
if (some_syscall() == EINTR) { ...
```

would become:

```
if (strnmp(some_syscall(), "interrupted") == 0) { ...
```

An then you commit to the exact string format anyway...

Reply:
It depends if the consumer is ever going to want to take an different actions based on the returned error...

Reply:
We don't currently expect the code to *recover* from an error -- we just want them to know why we failed. In general "you don't have an Intel PT capable processor" is something the end user has to fix, not the programmer!

Reply:
We can move forward with a free-form string for now, but as soon as clients need to differentiate errors, I think we will need to rethink it so that they wouldn't have to do string comparisons.

Reply:
Fixed in https://github.com/ykjit/yk/pull/632/commits/36a1a0c11bc98ae54cdf4ef6d92cb010c632d04b.